### PR TITLE
Add Go-based integration test framework using testscript

### PR DESCRIPTION
**Pull request by michaeldwan**
## Summary

Replaces Python/pytest-based integration tests with a new **Go testscript framework** that is faster, more maintainable, and easier to write. The framework includes **60 comprehensive tests** covering core predictor functionality, builds, training, subprocess behavior, and HTTP server testing.

## Cog CLI Change: BUILDKIT_PROGRESS Environment Variable Support

This PR includes a fix to the cog CLI to respect the `BUILDKIT_PROGRESS` environment variable for the `--progress` flag default. Previously the CLI always defaulted to `auto`, ignoring the env var.

**Change**: `pkg/cli/build.go` - Check `BUILDKIT_PROGRESS` env var before falling back to default value.

This allows users to set `BUILDKIT_PROGRESS=quiet` to suppress verbose Docker build output, which is used by the integration test harness to keep test logs clean.

### Problems with Python Tests
- **Slow**: Sequential execution, fixtures in separate directories
- **Flaky**: Shared fixture directories caused race conditions
- **Complex setup**: Requires tox, pytest, and Python environment
- **Hard to maintain**: Tests spread across multiple files and directories

### Benefits of Go Tests
- **Self-contained**: Each `.txtar` file embeds fixtures, code, and assertions
- **Fast**: Parallel execution with automatic Docker cleanup (10x faster)
- **No Python deps**: Runs with `go test`, builds cog binary on-demand
- **Easy to write**: Simple command script format, inline fixtures
- **Reliable**: Fresh temp directory per test, proper cleanup

### Test Framework (`integration-tests/`)
- **Harness** (`harness/harness.go`) - Binary resolution, Docker cleanup, custom commands
- **Test runner** (`suite_test.go`) - Main test runner with `[slow]` condition
- **60 test fixtures** (`tests/*.txtar`) - Self-contained test cases

#### Basic Types & I/O (27 tests)
- Primitives: string, int, File, Path
- Collections: lists, optional types
- Complex: BaseModel output, unions
- Python versions: 3.8, 3.12, 3.13
- Pydantic 1.x and 2.x
- Training workflows
- Error handling

#### Build Features (8 tests)
- System packages (apt, ffmpeg, zsh)
- Python packages (pip, local wheels)
- Build configurations
- Error cases

#### Advanced Features (7 tests)
- Secrets
- Config overrides
- Training with setup
- Fast build mode
- Migrations

#### Edge Cases (7 tests)
- Runtime type handling
- Complex file types (GLB)
- Async predictors
- Complex nested types

#### Framework/GPU Tests (7 tests) [slow]
- TensorFlow 2.11.1
- PyTorch 2.7.0 + CUDA 12.6
- PyTorch 2.7.1 + CUDA 12.8
- Base image variations

#### Subprocess/Server Tests (4 tests) [slow]
- ✅ `setup_subprocess_simple` - Subprocess with SIGUSR1 signals
- ✅ `setup_subprocess_double_fork` - Double fork daemonization
- ✅ `setup_subprocess_double_fork_http` - HTTP server in subprocess
- ✅ `setup_subprocess_multiprocessing` - Python multiprocessing

#### `cog` command
Runs any cog subcommand with special handling for:
- `cog serve` - Runs in background with health checks
- Other subcommands run normally (build, predict, etc.)
- Environment variable expansion
- Negation support with `!`

#### `curl` command
Makes HTTP requests to `cog serve`:
```txtar
curl POST /predictions '{"input":{"s":"hello"}}'
stdout '"output":'
```

#### `wait-for` command
Waits for conditions with timeout (for subprocess readiness):
```txtar
wait-for file .ready 60s
wait-for http http://127.0.0.1:7777/ping 200 60s
```

#### `retry-curl` command
HTTP requests with automatic retry logic:
```txtar
retry-curl POST /predictions '{"input":{"s":"test"}}' 10 1s
```

### Harness Features

- **On-demand builds**: Automatically builds cog from source if not provided
- **Binary caching**: Reuses built binary across test runs
- **Docker cleanup**: Removes test images automatically
- **Port allocation**: Random ports for `cog serve` to avoid conflicts
- **Health checks**: Waits for server readiness before testing
- **Environment isolation**: Fresh temp dir per test
- **Parallel execution**: Tests run concurrently with throttling
- **Slow test marking**: Use `[slow]` to skip with `COG_TEST_FAST=1`
- **Quiet build output**: Uses `BUILDKIT_PROGRESS=quiet` to suppress Docker build noise

## Usage

```bash
# Run all tests (including slow GPU/framework tests)
make test-integration-go

# Run fast tests only (skip slow tests)
COG_TEST_FAST=1 make test-integration-go

# Run specific test
cd integration-tests && go test -v -run TestIntegration/string_predictor

# Run with custom cog binary
COG_BINARY=/path/to/cog make test-integration-go

# Run with specific Go test flags
cd integration-tests && go test -v -timeout 30m -run TestIntegration
```

## Writing Tests

Tests are `.txtar` files with embedded fixtures:

### Basic Test
```txtar
# Test description
cog build -t $TEST_IMAGE
cog predict $TEST_IMAGE -i s=world
stdout 'hello world'

-- cog.yaml --
build:
  python_version: "3.12"
predict: "predict.py:Predictor"

-- predict.py --
from cog import BasePredictor

class Predictor(BasePredictor):
    def predict(self, s: str) -> str:
        return "hello " + s
```

### Server Test
```txtar
cog build -t $TEST_IMAGE
cog serve
curl POST /predictions '{"input":{"s":"test"}}'
stdout '"output":"hello test"'

-- cog.yaml --
# ... same as above
```

### Subprocess Test with Readiness Signal
```txtar
cog build -t $TEST_IMAGE
cog serve
wait-for file .ready 60s
retry-curl POST /predictions '{"input":{"s":"test"}}' 10 500ms
stdout '"output":"hello test"'
```

### Slow Test (GPU/Framework)
```txtar
[slow] skip 'requires GPU/long build time'

cog build -t $TEST_IMAGE
cog predict $TEST_IMAGE
stdout 'result'

-- cog.yaml --
build:
  python_version: "3.12"
  gpu: true
  python_packages:
    - torch==2.7.0
```

## Python Tests Cleanup

Removed **43 Python tests** now covered by Go:
- ✅ Deleted `test_migrate.py` (3 tests)
- ✅ Deleted `test_train.py` (3 tests)
- ✅ Removed 39 fixture directories
- ✅ Removed 12 tests from `test_build.py`
- ✅ Removed 25 tests from `test_predict.py`

**37 Python tests remain** for CLI flags and tooling:
- `--json` flag (5 tests)
- `-o` output flag (2 tests)
- `cog run` command (7 tests)
- `cog init` / `cog install` (2 tests)
- Config validation, TTY handling, etc.

## Test Coverage Philosophy

The two test suites are **complementary**:

### Go Tests (60) - Core Functionality ✅
- ✅ All predictor types and I/O patterns
- ✅ Build configurations and dependencies
- ✅ Training workflows
- ✅ Subprocess behavior
- ✅ HTTP server testing
- ✅ Framework integration (PyTorch, TensorFlow)
- ✅ Error handling

### Python Tests (37) - CLI & Tooling
- CLI flags not in Go tests yet
- `cog run` command (not in Go tests)
- Complex stdin/stdout/TTY interactions
- Config validation error messages

### Key Commits
1. Add Go testscript framework foundation
2. Port 27 basic type tests
3. Port 13 additional tests (builds, training)
4. Port 22 more tests (advanced features, edge cases)
5. Add framework/GPU tests with `[slow]` marking
6. Add HTTP server testing (`cog serve`, `curl`)
7. Port subprocess tests
8. Remove obsolete Python tests
9. Documentation updates

### Improvements Made
- ✅ Fix `COG_BINARY` resolution for relative paths
- ✅ Support `BUILDKIT_PROGRESS` env var in CLI for quiet builds
- ✅ Increase parallelism for faster test runs
- ✅ Add `[slow]` condition for test filtering
- ✅ Add `wait-for` and `retry-curl` commands for reliable subprocess testing

## Migration Status

| Category | Status | Tests |
|----------|--------|-------|
| Basic types & I/O | ✅ Complete | 27 |
| Builds | ✅ Complete | 8 |
| Advanced features | ✅ Complete | 7 |
| Edge cases | ✅ Complete | 7 |
| Frameworks/GPU | ✅ Complete | 7 |
| Subprocess/server | ✅ Complete | 4 |
| CLI flags | 🔄 Python only | 7 |
| `cog run` | 🔄 Python only | 7 |
| Other tooling | 🔄 Python only | 23 |

**Total: 60 Go tests + 37 Python tests = 97 integration tests**

### CODE REVIEW

File: integration-tests/tests/setup_subprocess_multiprocessing.txtar [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+# This ensures stream redirection works correctly with Python multiprocessing
+
+# Build the image
+cog build -t $TEST_IMAGE
+
+# Start the server
+cog serve
+
+# Wait for setup to complete (multiprocessing initialization takes time)
+exec sleep 2
```
Comment (markphelps) [inline integration-tests/tests/setup_subprocess_multiprocessing.txtar]: i dont love the 'sleeps' here. makes the tests flakey. we need true healtcheck support

**Comment by michaeldwan** [comment]
@markphelps nice work! would you be able to add a readme to the integration tests dir that covers how to run these (command, filtering, etc), how to add more, callout the editor extension for syntax highlights (my brain can't handle these as plain text files 🙈) and whatnot. just something contributors/agents can read to understand how to get the most out of this
